### ISSUE TITLE
Request allocation strategy

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2484]
Allocating from the std allocator means some inefficiency, plus a lot of work during destruction if the data involved a lot of small objects.

We could introduce a request-oriented `allocation_strategy`, with the following features:

 - bump-the-pointer allocation (like lsa)
 - multiple discontiguous memory segments (like lsa)
 - fragmented managed_bytes (like lsa)
 - exponentially growing segments, up to a limit (like bytes_ostream)
 - no-op free() (like Java)

This would be really fast, wasting some memory if we have lots of tombstones or overwrites, but I think this is acceptable.

If/when we implement #1865 then this will be more long-lived and so we do need some freeing. I think that could be done with something akin to replay_position:

 - a rows_entry that is allocated acquires a replay_position (refcount into the segments that it has memory in)
 - when it is destroyed, it releases the ref
 - when all refs are gone, the segment is freed

so mutation_readers would allocate new segments, and as fragments are consumed, segments would be freed as all mutation_fragments referring to them are discarded.

**Comment by pdziepak** [comment]
I was experimenting with something similar [1]. However, I postponed (abandoned?) it in favour of the new in-memory representation which should deal with some of the problems in a better way (namely: reduce the number of allocated objects and limit the size of a single object).

[1] https://gist.github.com/pdziepak/92417a021e6a6051ff5ea2fb8661060e

**Comment by avikivity** [comment]
We'd need to drop the requirement for linearized managed_bytes for this to work; we could do this using double-compiling the code that touches managed_bytes with char* and double-deref iterator.

**Comment by avikivity** [comment]
@pdziepak I'm worried about the runtime costs of apply() using the new imr. I think that imr is better in cache where reducing object size has a lot of value, compared to streamed mutations where cpu cost and ease of manipulation is more important.

**Comment by avikivity** [comment]
```c++
template <typename Func>
with_managed_bytes_view(const managed_bytes& b, Func&& func) {
    if (b.is_linear()) {
        func(make_iterator_range(b.simple_begin(), b.simple_end()));
    } else {
        func(make_iterator_range(b.complicated_begin(), b.complicated_end());
    }
}
```

With a variant for multiple manage_bytes objects.

**Comment by pdziepak** [comment]
A long term solution for reducing the impact of data manipulation could be to use expression templates. E.g. if we have a read path like this:
```
merge(sstables | filter(belong_to_this_shard), memtables)
     | compact
     | data_query
```
We could make sure that there all applies and the compaction are squashed into just a single operation so that the buffers are reallocated only once (or not at all if data query is squashed with that as well, but that may be trickier).

A shorter term solution could be to just use a more memory wasteful representation that is more likely to allow operation to be performed in place (switching implementation is not an additional cost since we need a copy when moving from one allocator to another anyway), while still having all the nice properties (limit of maximum object size, avoiding indirections). IMR should make it easier to have multiple representations.

**Comment by avikivity** [comment]
Expression templates are definitely a good fit here!

**Comment by avikivity** [comment]
But: combining_reader cannot be used there, just for 1:1 filtering.
